cdc/entry/schema_storage.go

```diff
@@ -678,79 +539,118 @@ func (s *Storage) CloneTables() map[uint64]TableName {
-		n.tables[k] = v.Clone()
+	return s.snaps[i-1], nil
+}
+
+// GetSnapshot returns the snapshot which of ts is specified
+func (s *SchemaStorage) GetSnapshot(ts uint64) (*schemaSnapshot, error) {
+	var snap *schemaSnapshot
+	err := retry.Run(10*time.Millisecond, 25,
+		func() error {
+			var err error
+			snap, err = s.getSnapshot(ts)
+			if errors.Cause(err) != model.ErrUnresolved {
+				return backoff.Permanent(err)
+			}
+			return err
+		})
+	switch err.(type) {
+	case *backoff.PermanentError:
+		return nil, errors.Annotate(err, "timeout")
+	default:
+		return snap, err
 	}
-	for k, v := range s.truncateTableID {
-		n.truncateTableID[k] = v
+}
+
+// GetLastSnapshot returns the last snapshot
+func (s *SchemaStorage) GetLastSnapshot() *schemaSnapshot {
+	s.snapsMu.RLock()
+	defer s.snapsMu.RUnlock()
+	return s.snaps[len(s.snaps)-1]
+}
+
+// HandleDDLJob creates a new snapshot in storage and handles the ddl job
+func (s *SchemaStorage) HandleDDLJob(job *timodel.Job) error {
+	s.snapsMu.Lock()
+	defer s.snapsMu.Unlock()
+	lastSnap := s.snaps[len(s.snaps)-1]
+	if job.BinlogInfo.FinishedTS <= lastSnap.currentTs {
+		log.Debug("ignore foregone DDL job", zap.Reflect("job", job))
+		return nil
 	}
-	for k, v := range s.ineligibleTableID {
-		n.ineligibleTableID[k] = v
+	if SkipJob(job) {
+		s.AdvanceResolvedTs(job.BinlogInfo.FinishedTS)
+		return nil
 	}
-	for k, v := range s.version2SchemaTable {
-		n.version2SchemaTable[k] = v
+	log.Info("handle ddl job", zap.Reflect("job", job))
+	snap := lastSnap.Clone()
+	if err := snap.handleDDL(job); err != nil {
+		return errors.Trace(err)
 	}
-	n.schemaMetaVersion = s.schemaMetaVersion
-	n.lastHandledTs = s.lastHandledTs
-	n.resolvedTs = s.resolvedTs
-	n.jobList = s.jobList
-	n.currentJob = s.currentJob
-	n.currentVersion = s.currentVersion
-	return n
-}
-
-// IsTruncateTableID returns true if the table id have been truncated by truncate table DDL
-func (s *Storage) IsTruncateTableID(id int64) bool {
-	_, ok := s.truncateTableID[id]
-	return ok
+	s.snaps = append(s.snaps, snap)
+	s.AdvanceResolvedTs(job.BinlogInfo.FinishedTS)
+	return nil
 }
 
-// IsIneligibleTableID returns true if the table is ineligible
-func (s *Storage) IsIneligibleTableID(id int64) bool {
-	_, ok := s.ineligibleTableID[id]
-	return ok
+// AdvanceResolvedTs advances the resolved
+func (s *SchemaStorage) AdvanceResolvedTs(ts uint64) {
+	atomic.StoreUint64(&s.resolvedTs, ts)
 }
 
-// FillSchemaName fills the schema name in ddl job
-func (s *Storage) FillSchemaName(job *timodel.Job) error {
-	var schemaName string
-	if job.Type != timodel.ActionCreateSchema {
-		dbInfo, exist := s.SchemaByID(job.SchemaID)
-		if !exist {
-			return errors.NotFoundf("schema %d not found", job.SchemaID)
+// DoGC removes snaps which of ts less than this specified ts
+func (s *SchemaStorage) DoGC(ts uint64) {
+	s.snapsMu.Lock()
+	defer s.snapsMu.Unlock()
+	var startIdx int
+	for i, snap := range s.snaps {
+		if snap.currentTs > ts {
```

Comment:
Do we need to keep snapshot with `snap.currentTs = ts`

Reply:
I don't think we need to consider this. One more snapshot does not affect anything, but one less snapshot does not work. So we leave one more